man/chsh.1
.TH CHSH 1 "1 Feb 1988" "Project Athena"
\" RCSID: $HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/man/chsh.1 $ $Id: chsh.1 3982 2010-02-16 21:15:17Z zacheiss $
.SH NAME
chsh \- change user login shell
.SH SYNOPSIS
.B chsh
[
.BI \-u
|
.BI \-w
] [
.BI \-s
.I shell
] [
.BI \-db
.I host:port
]
[ username ]
.SH DESCRIPTION
This command changes the login shell associated with
.IR username
(your own name by default).
While you can specify any program as a command interpreter, be
cautious about specifying a program that won't let you change your
login shell to something else (i.e., a noninteractive program like
"date").
.PP
When invoked,
.B chsh
displays the current shell, and prompts for the new one unless the
.BI \-s
option has been specified.
.PP
If invoked with the 
.BI \-w 
option, chsh will change the user's Windows shell instead of their
UNIX shell.  The default behavior is to change the user's UNIX shell,
and the 
.BI \-u
option is provided to explicitly select this behavior.

The actual change will take place following the next Hesiod update
from Moira, which generally occurs within three hours.
Users may only change their own information unless they are
Moira database administrators.
.SH AUTHOR
Emanuel 'Jay' Berkenbilt
.SH "SEE ALSO"
moira(1),
Moira section of the Project Athena Technical Plan


debian/changelog
debathena-moira (4.0.0+svn20100225-0debathena2) UNRELEASED; urgency=low

  * There is no "lib" section - it's "libs".

 -- Evan Broder <broder@mit.edu>  Fri, 26 Feb 2010 12:50:19 -0500

debathena-moira (4.0.0+svn20100225-0debathena1) unstable; urgency=low

  * New SVN snapshot.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 15:54:26 -0500

debathena-moira (4.0.0+svn20100223-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Drop patch install-headers, which was incorporated upstream.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 02:13:44 -0500

debathena-moira (4.0.0+svn20100104-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Build and install libmrclient as a shared library.
  * Deal with rename of /etc{/athena => }/moira.conf
  * Patch manpages to deal with our installation paths.

 -- Evan Broder <broder@mit.edu>  Mon, 04 Jan 2010 22:05:10 -0500

debathena-moira (4.0.0+cvs20091116-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop patch fix-update-ticket, which was incorporated upstream.
  * Add a missing definition of ${prefix} in include/Makefile.in so that
    the moira.h header actually gets installed in
    /usr/include/moira. (Trac: #445)

 -- Evan Broder <broder@mit.edu>  Thu, 19 Nov 2009 10:53:40 -0500

debathena-moira (4.0.0+cvs20090922-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Add patch to fix builds without krb4.

 -- Evan Broder <broder@mit.edu>  Tue, 22 Sep 2009 00:35:12 -0400

debathena-moira (4.0.0+cvs20090728-0debathena3) unstable; urgency=low

  * Add a missing conflicts to make upgrades run more smoothly.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 18:28:33 -0700

debathena-moira (4.0.0+cvs20090728-0debathena2) unstable; urgency=low

  * Include the headers in debathena-libmoira-dev.
  * Install update_test in moira-clients instead of moira-update-server.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 17:54:34 -0700

debathena-moira (4.0.0+cvs20090728-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Fix zephyr dependencies to actually pull in Kerberized libraries; drop
    build-dep on Kerberized library.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 13:43:28 -0400

debathena-moira (4.0.0+cvs20090621-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop krb4less-fix; it was incorporated upstream.
  * Universally stop building Moira with krb4 support.

 -- Evan Broder <broder@mit.edu>  Fri, 26 Jun 2009 00:51:28 -0400

debathena-moira (4.0.0+cvs20090618-0debathena1) unstable; urgency=low

  * New upstream release
  * Build and install libmoira as a shared library. (Trac: #70)
  * Drop most of the krb4 patch - it's been incorporated upstream.
  * Install the Moira development headers by patching the relevant
    Makefiles, instead of in the debian/rules file.

 -- Evan Broder <broder@mit.edu>  Sat, 20 Jun 2009 14:45:32 -0400

debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


clients/chsh/chsh.c
/* $Id: chsh.c 3981 2010-02-16 21:13:05Z zacheiss $
 *
 * Talk to the Moira database to change a person's login shell.  The chosen
 * shell must exist.  A warning will be issued if the shell is not in
 * /etc/shells.
 *
 * chsh with no modifiers changes the shell of the user who is running
 * the program.
 *
 * If a commandline argument is given, it is taken to be the username
 * of the user whose shell is to be changed.
 *
 * Copyright (C) 1988-1998 by the Massachusetts Institute of Technology
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 */

#include <mit-copyright.h>
#include <moira.h>
#include <moira_site.h>
#include <mrclient.h>

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/param.h>

#define argis(a, b) (!strcmp(*arg + 1, a) || !strcmp(*arg + 1, b))

RCSID("$HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/clients/chsh/chsh.c $ $Id: chsh.c 3981 2010-02-16 21:13:05Z zacheiss $");

void usage(void);
int get_shell(int argc, char **argv, void *username);
int get_winshell(int argc, char **argv, void *username);
int get_fmodtime(int argc, char **argv, void *username);
void check_shell(char *shell);
#ifndef HAVE_GETUSERSHELL
char *getusershell(void);
#endif

char *whoami;
char *username;

int main(int argc, char *argv[])
{
  int status;                   /* general purpose exit status */
  int q_argc;                   /* argc for mr_query */
  char *q_argv[U_END];          /* argv for mr_query */
  char *motd;                   /* determine Moira server status */
  int got_one = 0;              /* Have we got a shell yet? */
  int windowsflg = 0, unixflg = 0;  
  char **arg = argv;
  char *server = NULL, *shell = NULL;

  if ((whoami = strrchr(argv[0], '/')) == NULL)
    whoami = argv[0];
  else
    whoami++;

  /* parse our command line options */
  while (++arg - argv < argc)
    {
      if (**arg == '-')
	{
	  if (argis("w", "winshell"))
	    windowsflg++;
	  else if (argis("u", "unixshell"))
	    unixflg++;
	  else if (argis("s", "shell"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  shell = *arg;
		}
	      else
		usage();
	    }
	  else if (argis("db", "database"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  server = *arg;
		}
	      else
		usage();
	    }
	}
      else if (username == NULL)
	username = *arg;
      else
	usage();
    }

  if (!username)
    {
      username = mrcl_krb_user();
      if (!username)
	exit(1);
    }

  if (!unixflg && !windowsflg)
    unixflg++;

  if (unixflg && windowsflg)
    usage();
  
  if (mrcl_connect(server, "chsh", 14, 1) != MRCL_SUCCESS)
    exit(1);

  /* First, do an access check */

  q_argv[USH_NAME] = username;
  q_argv[USH_SHELL] = "junk";
  q_argc = USH_END;

  if ((status = mr_access("update_user_shell", q_argc, q_argv)))
    {
      com_err(whoami, status, "; shell not\nchanged.");
      exit(2);
    }

  printf("Changing login shell for %s.\n", username);

  /* Display current information */
  
  q_argv[NAME] = username;
  q_argc = NAME + 1;
  
  if ((status = mr_query("get_finger_by_login", q_argc, q_argv,
			 get_fmodtime, username)))
    {
      com_err(whoami, status, "while getting user information.");
      exit(2);
    }

  if (unixflg)
    {
      if ((status = mr_query("get_user_account_by_login", q_argc, q_argv,
			     get_shell, username)))
	{
	  com_err(whoami, status, "while getting user information.");
	  exit(2);
	}


      if (!shell)
	{
	  /* Ask for new shell */
	  while (!got_one)
	    {
	      shell = malloc(MAXPATHLEN);
	      if (!shell)
		exit(1);
	      printf("New shell: ");
	      if (!fgets(shell, MAXPATHLEN, stdin))
		exit(0);
	      got_one = (strlen(shell) > 1);
	    }

	  shell[strlen(shell) - 1] = 0; /* trim newline */
	}

      /* Make sure we have a valid shell.  This routine could exit */
      check_shell(shell);
      
      /* Change shell */
      
      printf("Changing shell to %s...\n", shell);
      
      q_argv[USH_NAME] = username;
      q_argv[USH_SHELL] = shell;
      q_argc = USH_END;
      if ((status = mr_query("update_user_shell", q_argc, q_argv, NULL, NULL)))
	{
	  com_err(whoami, status, "while changing shell.");
	  exit(2);
	}
      
      printf("Shell successfully changed.\n");
    }
  else if (windowsflg)
    {
      if ((status = mr_query("get_user_account_by_login", q_argc, q_argv,
			     get_winshell, username)))
	{
	  com_err(whoami, status, "while getting user information.");
	  exit(2);
	}

      if (!shell)
	{
	  /* Ask for new Windows shell */
	  while(!got_one)
	    {
	      shell = malloc(MAXPATHLEN);
	      if (!shell)
		exit(1);
	      printf("New Windows shell: ");
	      if (!fgets(shell, MAXPATHLEN, stdin))
		exit(0);
	      got_one = (strlen(shell) > 1);
	    }
      
	  shell[strlen(shell) - 1] = 0; /* strip off newline */
	}

      /* Change shell */
      
      printf("Changing Windows shell to %s...\n", shell);
      
      q_argv[USH_NAME] = username;
      q_argv[USH_SHELL] = shell;
      q_argc = USH_END;
      if ((status = mr_query("update_user_windows_shell", q_argc, q_argv,
			     NULL, NULL)))
	{
	  com_err(whoami, status, "while changing Windows shell.");
	  exit(2);
	}
      
      printf("Windows shell successfully changed.\n");
    }
     
  mr_disconnect();
      
  return 0;
}

int get_shell(int argc, char **argv, void *username)
{
  /* We'll just take the first information we get since login names
     cannot be duplicated in the database. */

  if (argc < U_END || strcmp(argv[U_NAME], username))
    {
      fprintf(stderr, "Some internal error has occurred.  Try again.\n");
      exit(3);
    }

  printf("Current shell for %s is %s.\n", (char *)username, argv[U_SHELL]);

  return MR_ABORT;		/* Don't pay attention to other matches. */
}

int get_winshell(int argc, char **argv, void *username)
{
  /* We'll just take the first information we get since login names
     cannot be duplicated in the database. */

  if (argc < U_END || strcmp(argv[U_NAME], username))
    {
      fprintf(stderr, "Some internal error has occurred.  Try again.\n");
      exit(3);
    }

  printf("Current Windows shell for %s is %s.\n", (char *)username, 
	 argv[U_WINCONSOLESHELL]);

  return MR_ABORT;		/* Don't pay attention to other matches. */
}

int get_fmodtime(int argc, char **argv, void *username)
{
  /* We'll just take the first information we get since login names
     cannot be duplicated in the database. */

  if (argc < F_END || strcmp(argv[F_NAME], username))
    {
      fprintf(stderr, "Some internal error has occurred.  Try again.\n");
      exit(3);
    }

  printf("Finger information last changed on %s\n", argv[F_MODTIME]);
  printf("by user %s with %s.\n", argv[F_MODBY], argv[F_MODWITH]);

  return MR_ABORT;		/* Don't pay attention to other matches. */
}

void check_shell(char *shell)
{
  char *valid_shell;
  int ok = 0;

  while ((valid_shell = (char *)getusershell()))
    {
      if (!strcmp(shell, valid_shell))
	{
	  ok = 1;
	  break;
	}
      else if (!strcmp(shell, 1 + strrchr(valid_shell, '/')))
	{
	  ok = 1;
	  strcpy(shell, valid_shell);
	  break;
	}
    }

  if (!ok)
    {
      if (shell[0] != '/')
	{
	  fprintf(stderr, "%s is not a standard shell.  ", shell);
	  fprintf(stderr, "You may choose to use a nonstandard\n");
	  fprintf(stderr, "shell, but you must specify its complete ");
	  fprintf(stderr, "path name.\n");
	  exit(2);
	}
      else if (access(shell, X_OK))
	{
	  fprintf(stderr, "%s is not available.\n", shell);
	  exit(2);
	}
      else
	{
	  printf("%s exists but is an unusual choice.\n", shell);
	  printf("Try again if it is not what you wanted.\n");
	}
    }
}

void usage(void)
{
  fprintf(stderr, "Usage: %s [-w|-u] [-s shell] [user]\n", whoami);
  exit(1);
}

#ifndef HAVE_GETUSERSHELL
#include <sys/param.h>

char *getusershell(void)
{
  static FILE *shells = NULL;

  /* In a sane universe, no shell will have a length longer than
   * MAXPATHLEN. If any line in /etc/shells does, we'll lose, but
   * not much. shrug.
   */
  static char buf[MAXPATHLEN];
  char *p;

  if (!shells)
    {
      shells = fopen("/etc/shells", "r");
      if (!shells)
	{
	  fprintf(stderr, "%s: Can't open /etc/shells. Unable to determine if "
		  "this is a normal shell.\n\n", whoami);
	  return NULL;
	}
    }

  while (1)
    {
      if (!fgets(buf, sizeof(buf), shells))
	{
	  fclose(shells);
	  shells = NULL;
	  return NULL;
	}

      if (buf[0] != '/')
	continue;

      p = strchr(buf, '\n');
      if (p)
	*p = '\0';

      return buf;
    }
}
#endif
